packages/lesswrong/components/posts/BookmarkButton.tsx

```diff
@@ -27,17 +28,26 @@ const BookmarkButton = ({classes, post, menuItem, placement="right"}: {
 }) => {
   const currentUser = useCurrentUser();
   const { openDialog } = useDialog();
-  const [bookmarked, setBookmarked] = useState(_.pluck((currentUser?.bookmarkedPostsMetadata || []), 'postId')?.includes(post._id))
-  const { captureEvent } = useTracking({eventType: "bookmarkToggle", eventProps: {"postId": post._id, "bookmarked": !bookmarked}})
+  const [bookmarked, setBookmarkedState] = useState(_.pluck((currentUser?.bookmarkedPostsMetadata || []), 'postId')?.includes(post._id))
```

Comment:
Just for future reference, I would just solve this with optimistic response. I don't like having this profusion of parallel state-setting mechanisms, and I've had no problem doing stuff with optimistic response in the past (it worked really well for the Elicit Block widget, for example).

Reply:
The primary reason for my distaste is just that it fails really unelegantly in the case of errors.

Reply:
(This is preexisting code; I renamed the setter to avoid a name conflict but did not otherwise change the client-state mechanism.)